builtin/game/item_entity.lua

```diff
@@ -318,15 +318,49 @@ core.register_entity(":__builtin:item", {
 		end
 	end,
 
-	on_punch = function(self, hitter)
-		local inv = hitter:get_inventory()
-		if inv and self.itemstring ~= "" then
-			local left = inv:add_item("main", self.itemstring)
-			if left and not left:is_empty() then
-				self:set_item(left)
+	on_punch = function(self, hitter, ...)
+		if self.itemstring ~= "" then
+			-- Call on_pickup callback in item definition.
+			local itemstack = ItemStack(self.itemstring)
+			local itemstack_modified = false
+			local item_def = itemstack:get_definition()
+			if item_def and item_def.on_pickup then
+				local ret = item_def.on_pickup(itemstack, hitter, self.object, ...)
+				if not ret then
+					return
+				elseif ret ~= true then
+					itemstack_modified = true
+					itemstack = ret
```

Comment:
```suggestion
					itemstack = ItemStack(ret)
```
Untested, but should allow modders to return strings which are passed down to the next function as ItemStack.

Relevant: register.lua / `core.on_craft`.